en/lessons/specifics/benchmarking.md

```diff
@@ -0,0 +1,346 @@
+
+The console output that we've seen is a helpful beginning for measuring the
+runtimes of your functions, but it's not your only option! In this section we'll
+look briefly at the three other available formatters, and also touch on what
+you'd need to do to write your own formatter if you like.
+
+## Other formatters
+
+Benchee has a console formatter built in, which is what we've seen already, but
+there are three other officially supported formatters - `benchee_csv`,
+`benchee_json` and `benchee_html`. Each of them does exactly what you would
+expect, which is writing the results to the named file formats so you can work
+with your results further in whichever format you like.
+
+Each of these formatters is a separate package, so to use them you need to add
+them as dependencies to your `mix.exs` file like so:
+
+```elixir
+defp deps do
+  [
+    {:benchee_csv,  "~> 0.6", only: :dev},
+    {:benchee_json, "~> 0.3", only: :dev},
+    {:benchee_html, "~> 0.3", only: :dev},
+  ]
+end
+```
+
+While `benchee_json` and `benchee_csv` are very simple, `benchee_html` is
+actually _very_ full featured! It can help you produce nice graphs and charts of
+from your results easily, and you can even export them as PNG images. All three
+formatters are well documented on their respective GitHub pages, so we won't
+cover the details of them here.
+
+## Custom formatters
+
+If the four offered formatters aren't enough for you, you can also write your own formatter. Writing a formatter is pretty easy. You need to write a function that accepts a `%Benchee.Suite{}` struct, and from that you can pull whatever information you like. Information about what exactly is in this struct can be found on [GitHub](https://github.com/PragTob/benchee/blob/master/lib/benchee/suite.ex) or [HexDocs](https://hexdocs.pm/benchee/Benchee.Suite.html). The codebase is very well documented and easy to read if you'd like to see what sorts of information could be available for writing custom formatters.
+
+For now, I'll show a quick example of what a custom formatter might look like below as an example of how easy it is. Let's say we just want an extremely minimal formatter that just prints the average run time for each scenario - this is what that might look like:
+
+```elixir
+defmodule Custom.Formatter do
+  def output(suite) do
+    suite
+    |> format
+    |> IO.write
+
+    suite
+  end
+
+  defp format(suite) do
+    Enum.map_join(suite.scenarios, "\n", fn(scenario) ->
+      "Average for #{scenario.job_name}: #{scenario.run_time_statistics.average}"
+    end)
+  end
+end
+```
+
+And then we could run our benchmark like this:
+
+```elixir
+list = Enum.to_list(1..10_000)
+map_fun = fn(i) -> [i, i * i] end
+
+Benchee.run(%{
+  "flat_map"    => fn -> Enum.flat_map(list, map_fun) end,
+  "map.flatten" => fn -> list |> Enum.map(map_fun) |> List.flatten end
+}, formatters: [&Custom.Formatter.output/1])
+```
+
+And when we run now with our custom formatter, we would see:
+
+```shell
```

Comment:
Maybe we shouldn't specify the language for this section, because `for` will be recognised as a keyword.

Reply:
I added this because it explicitly mentioned using `shell` in the style guide. If you think it makes sense to go against the style guide in this example, then I'm all for it. I didn't originally read this comment as an ok to do that, more of a question. I'm happy either way.